### DISCUSSION TITLE
Dynamic reading and writing build problems

### DISCUSSION THREAD

**Discussion by clave-sounds** [melanchall/drywetmidi#264] [General]
Hi Max. 

As you may remember, my project is dynamically writing and then reading a single midiFile. 

It works great in unity, but when I build for Android, it gets stuck and I think the problem is at the point where it either writes or reads the midiFile. I've tried using Application.persistentDataPath and Application.temporaryCachePath and both still get stuck. Should I be using a different file path to make this work? 

Any thoughts on what might be happening? 

Thanks

**Comment by melanchall** [comment]
Hi,

Show the code please.

Also clarify what do you mean _"it gets stuck"_? Is Unity hanging?

**Comment by clave-sounds** [comment]
Hi Max. 

Ok, here's the code. I've tried to remove anything irrelevant so it's not miles of code. Please forgive messy code, this is my first project.



I hope that all makes sense. 

Thanks

**Reply by melanchall** [reply]
Hi,

Thanks for the info. Please write a problem more carefully in future issues/discussions. Because

> it gets stuck and I think the problem is at the point where it either writes or reads the midiFile

doesn't not equal to

> When I say Unity gets stuck, what I mean specifically is that when the play() method is called

Reading/writing is not playing. Well, your problem is clear now. You can't use Playback API on Android. Please read [Supported OS](https://melanchall.github.io/drywetmidi/articles/dev/Supported-OS.html) article.

Thanks,
Max

**Comment by clave-sounds** [comment]
Hi Max. Ok, apologies for my bad use of terminology. I'm still learning. I see what you mean, reading is not the same as playback, yes.

I hope you don't mind me asking a question unrelated to DryWetMidi, but I don't really have anyone else to get advice from that understand both music and coding. But am I using the wrong platform or language? I learned C#, but it seems impossible to do what I need. Unity seems designed to do good graphics, but it can't hold time for music. And yet, there are all these apps like iReal that do hold good time, so it must be possible. Do I need to learn swift and Kotlin? What would you recommend for making something that can sequence music, either straight from a tick Generator or via dynamic MidiFile? 
I've tried making my own Tick Generators with coRoutines and unity timer, but they aren't good, the time is unreliable. If it is possible with unity/C# what do I need to research to do it? 

As always, thanks for your time.

**Reply by melanchall** [reply]
Hi,

It's a pretty hard question. I suppose that .NET/C# is a good choice. The language is mature, there are a lot of different libs for .NET, it has large community and C# is one of the top languages many years. Swift in my opinion is a great for Apple systems. As for Kotlin – good language too.

As for tick generator, you can use DryWetMIDI with a custom one (see example here – [Custom tick generator](https://melanchall.github.io/drywetmidi/articles/playback/Tick-generator.html#custom-tick-generator)).

As for me, I use C# in my personal projects and on my main job, so I suppose I'm a bad adviser :)

**Comment by clave-sounds** [comment]
Oh that's interesting. Thanks Max. 

Actually since I wrote that last comment, I've had a bit more luck creating a tick generator by doing it based on the audio rate and using OnAudioFilterRead. This seems to hold time pretty well although the processor struggles depending on what I'm doing with the ticks. I've been finding ways to reduce that by putting events in dictionaries with the tick as the key so it doesn't have to iterate through the whole list every tick. It also helps if I only have a single event publisher.  

I think the bit that I was missing is the threading. If you want to create accurate pulses it has to be off the main thread. Because I notice that in your gen you are just using stopwatch, and I previously tried that on the main thread and it was really wobbly. 

I'll give your generator a proper try, maybe it will create accurate ticks without hammering the processor as much as the OnAudioFilterRead method does.

As always. Thanks for your advice and help. I can't stress how helpful it is to ask questions to someone that understands audio/Midi. Most unity devs think using sound means audioSource.play()  ;-)
